crates/proof-of-sql/src/sql/proof_exprs/range_check.rs

```diff
@@ -0,0 +1,382 @@
+                0, 0, 0,
+            ]
+        );
+
+        assert_eq!(
+            byte_rows[5],
+            &vec![
+                1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                0, 0, 0,
+            ]
+        );
+    }
+
+    #[test]
+    fn we_can_decompose_large_scalars_to_words() {
+        let mut scalars: Vec<S> = [S::MAX_SIGNED, S::from(u64::MAX), S::from(-1)]
+            .iter()
+            .map(S::from)
+            .collect();
+
+        let alloc = Bump::new();
+        let mut word_columns: Vec<&mut [u8]> = (0..31)
+            .map(|_| alloc.alloc_slice_fill_with(scalars.len(), |_| 0u8))
+            .collect();
+
+        let byte_counts = alloc.alloc_slice_fill_with(256, |_| 0u64);
+        let mut byte_rows: Vec<&[u8]> = Vec::with_capacity(scalars.len());
+
+        decompose_scalar_to_words(
+            &mut scalars,
+            &alloc,
+            &mut word_columns,
+            byte_counts,
+            &mut byte_rows,
+        );
+
+        print_scalar_bytes_matrix(&byte_rows);
+
+        // Check that the decomposition columns are correct
+        assert_eq!(&word_columns[0], &[246, 255, 236]);
+        assert_eq!(&word_columns[1], &[233, 255, 211]);
+        assert_eq!(&word_columns[2], &[122, 255, 245]);
+        // ...
+        assert_eq!(&word_columns[15], &[10, 0, 20]);
+
+        // Word row of S::MAX_SIGNED
+        assert_eq!(
+            byte_rows[0],
+            &vec![
+                246, 233, 122, 46, 141, 49, 9, 44, 107, 206, 123, 81, 239, 124, 111, 10, 0, 0, 0,
+                0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            ]
+        );
+
+        // Word row of u64::MAX
+        assert_eq!(
+            byte_rows[1],
+            &vec![
+                255, 255, 255, 255, 255, 255, 255, 255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                0, 0, 0, 0, 0, 0, 0, 0, 0,
+            ]
+        );
+
+        // Word row of S::from(-1) aka P -1
+        assert_eq!(
+            byte_rows[2],
+            &vec![
+                236, 211, 245, 92, 26, 99, 18, 88, 214, 156, 247, 162, 222, 249, 222, 20, 0, 0, 0,
+                0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            ]
+        );
+    }
+
+    #[test]
+    fn we_can_obtain_a_logarithmic_derivative_from_a_small_scalar() {
+        let mut scalars: Vec<S> = [1, 2, 3, 255, 256, 257].iter().map(S::from).collect();
+
+        let alloc = Bump::new();
+        let mut word_columns: Vec<&mut [u8]> = (0..31)
+            .map(|_| alloc.alloc_slice_fill_with(scalars.len(), |_| 0u8))
+            .collect();
+
+        let byte_counts = alloc.alloc_slice_fill_with(256, |_| 0u64);
+
+        let mut byte_rows: Vec<&[u8]> = Vec::with_capacity(scalars.len());
+
+        decompose_scalar_to_words(
+            &mut scalars,
+            &alloc,
+            &mut word_columns,
+            byte_counts,
+            &mut byte_rows,
+        );
```

Comment:
decompose_scalar_to_words gives the words that I need before computing a log derivative. Instead we can just put in some bytes by hand. it would be nice to see these both working and connected to each other at some point though